src/tools/time_governor.cc

```diff
@@ -364,20 +422,88 @@ void TimeGovernor::init_common(double init_time, double end_time, TimeMark::Type
 
 
 
-void TimeGovernor::set_permanent_constraint( double min_dt, double max_dt)
+void TimeGovernor::set_dt_limits( double min_dt, double max_dt)
 {
-    if (min_dt < time_step_precision) {
-		THROW(ExcTimeGovernorMessage()	<< EI_Message("'min_dt' smaller then machine precision.\n") );
+	dt_limits_table_.clear();
+
+	if (min_dt < time_step_precision) {
+		THROW(ExcTimeGovernorMessage()	<< EI_Message("'min_dt' smaller than machine precision.\n") );
     }
     if (max_dt < min_dt) {
-		THROW(ExcTimeGovernorMessage()	<< EI_Message("'max_dt' smaller then 'min_dt'.\n") );
+		THROW(ExcTimeGovernorMessage()	<< EI_Message("'max_dt' smaller than 'min_dt'.\n") );
     }
 
-    lower_constraint_ = min_time_step_ = max(min_dt, time_step_precision);
+    // fill table limits with two records (start time, end time)
+    dt_limits_table_.push_back( DtLimitRow(init_time_, min_dt, max_dt) );
+    dt_limits_table_.push_back( DtLimitRow(end_time_, min_dt, max_dt) );
+    dt_limits_iter_ = dt_limits_table_.begin();
+
+    set_permanent_constraint();
+}
+
+
+void TimeGovernor::set_dt_limits( double min_dt, double max_dt, Input::Array dt_limits_list)
```

Comment:
Try to merge these two functions, pass an empty array as a default value for the corresponding input key.